doc/rfc001-sum-types.org

```diff
@@ -0,0 +1,858 @@
+#+TITLE: RFC 001: Extend Quint with Row-Polymorphic Sum Types
+#+AUTHOR: Shon Feder
+#+LATEX_COMPILER: xelatex
+
+* Overview
+This section gives a concise overview of the principle proposal. In
+depth discussion follows in the subsequent sections.
+
+** Concrete syntax
+Here I illustrate the concrete syntax additions being proposed. See [[Concrete Syntax][the
+discussion of concrete syntax]] for consideration of motivations and trade offs.
+
+*** Declaration
+The proposed syntax follows fits with our existing syntax for records
+and takes inspiration from Rust.
+
+#+begin_src quint
+type T = {
+  A(int),
+  B(str),
+  C,
+}
+#+end_src
+
+As indicated by =C=, labels with no associated expression type are
+allowed. These are sugar for a label typed with the unit (i.e., the
+empty record). The above will be normalized as
+
+#+begin_src quint
+type T = {
+  A(int),
+  B(str),
+  C({}),
+}
+#+end_src
+
+*** Constructors
+Following our existing practice of maintaining exotic operators as the
+normal form of data constructors, we may add an operator =variant=
+similar to the "injection" operator from
+[[https://www.microsoft.com/en-us/research/publication/extensible-records-with-scoped-labels/][Leijen05]]:
+
+#+begin_src haskell
+<l = _> :: ∀αr . α → <l :: α | r > -- injection
+#+end_src
+
+This operation injects an expression of type =α= into a sum-type.
+However, if we don't want to expose the row-polymorphism to users, we'll
+need a more restrictive typing. This is discussed with the typing rules.
+
+When a sum type declaration is parsed, constructor operations for each
+variant will be generated internally. E.g., for the type =T= defined
+above, we will add to the context the following operators:
+
+#+begin_src quint
+pure def A(x:int): T = variant("A", x)
+pure def B(x:str): T = variant("B", x)
+pure def C(): T      = variant("C", {})
```

Comment:
The latter would conflict with our module system notion: it would not be clear from syntax whether you were accessing a module operator.

If we supported nested modules, we could generate a module with operators for the constructors, but we don’t support that. So I don’t know how we could resolve that ambiguity. I also don’t see the benefit for practically to that syntax Do you have anything in mind?

Reply:
Afaict, Rust doesn’t have that ambiguity because their modules names must be lower case. function names must also be lower case, upper case is reserved for data constructors.